teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+---
+status: proposed
+title: Array result types
+creation-date: '2021-07-14'
+last-updated: '2021-07-14'
+authors:
+- '@bobcatfish'
+---
+
+# TEP-0076: Expanded array support: results and indexing
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats](#notescaveats)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Upgrade &amp; Migration Strategy](#upgrade--migration-strategy)
+- [Implementation Pull request(s)](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes expanding support for arrays by adding:
+* Support for indexing into arrays in variable replacement
+  (in addition to [replacing entire lists](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#substituting-array-parameters))
+* Support for array [results](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) (for both Pipelines and Tasks)
+
+## Motivation
+
+* See [use cases](#use-cases) - use cases for looping support in pipelines often need an array input;
+  being able to produce an array from a Task to feed into a loop would enable some interesting CD use cases
```

Comment:
I'm still not sold on having loop support in pipeline itself (aka as a custom task looks just fine to me).

Reply:
Obviously, it could be a custom tasks but I think this is such a powerful feature that it deserves to be in the core API. If for no other reason than once you add the dynamism of looping, you add an extra dimension which Tekton can be extended on. Once you have looping, you can now support trivial replication fan-out, or conditional looping with when expressions, etc all with existing primitives.

Reply:
Having native loop support in the core API does not mean that we cannot have a dedicated controller/CRD for it still.
Loops are probably a very good use case for TEP-0044, where we might not want to have one dedicated Pod for each iteration. A `TaskGroup` controller could decide what is the best `TaskRuns` <-> `Pod` ratio based on the inputs.

Reply:
we've got a TEP for this now! :D 🎉 https://github.com/tektoncd/community/pull/600